### PR TITLE
Spruce up deterministic_cache, ecosystem and eigen

### PR DESCRIPTION
**Pull request by langner**
Next batch for #347 

* Minor improvements to docstrings and unit tests for ecosystem
* Simplified unit tests for deterministic_cache a little bit more
* Minor edits to eigen module, and made several function internal

### CODE REVIEW

File: axelrod/ecosystem.py [hunk 1]
```diff
@@ -1,15 +1,47 @@
+"""Tools for simulating population dynamics of immutable players.
+
+An ecosystem runs in the context of a previous tournament, and takes the
+results as input. That means no matches are run by the ecosystem, and a
+tournament needs to happen before it is create. For example:
```
Comment (meatballs) [inline axelrod/ecosystem.py]: typo: create should be created

File: axelrod/ecosystem.py [hunk 2]
```diff
@@ -1,15 +1,47 @@
...
+
+players = [axelrod.Cooperator(), axlerod.Defector()]
+tournament = axelrod.Tournament(players=players)
+results = tournament.play()
+ecosystem = axelrod.Ecosystem(results)
+acosystem.reproduce(100)
```
Comment (meatballs) [inline axelrod/ecosystem.py]: typo: acosystem

**Comment by drvinceknight** [comment]
Appveyor is failing because of #1200, travis is failing because of #1202, I believe #1201 fixes both of those.

**Comment by drvinceknight** [comment]
If you could pull-rebase/merge on to master @langner those two failures should now be fixed :+1:

File: axelrod/ecosystem.py (hunk 1 above)
Comment (langner) [reply axelrod/ecosystem.py]: Done.

File: axelrod/ecosystem.py (hunk 2 above)
Comment (langner) [reply axelrod/ecosystem.py]: Done.

File: axelrod/ecosystem.py [hunk 3]
```diff
@@ -1,15 +1,47 @@
...
+    num_players: int
+        The number of players
+    """
 
     def __init__(self, results: ResultSet,
                  fitness: Callable[[float], float] = None,
                  population: List[int] = None) -> None:
+        """Create a new ecosystem.
+        
+        Paramters
```
[the remark this answers is not in the document]
Comment (langner) [reply axelrod/ecosystem.py]: Done.

**Comment by langner** [comment]
Thanks for fixing the seed thing, I wouldn't have figured it out that fast by myself! This should now be good to go.

**Comment by drvinceknight** [comment]
No problem at all :) 

This all looks good to me now, merge when you're happy @meatballs :+1:

**Pull request merged**